src/frontend/XAPI/XAPIStatement.ts

```diff
@@ -0,0 +1,391 @@
+      } else {
+        playedSegment.push(`${this.startSegments[i]}`);
+      }
+    }
+
+    return playedSegment.join('[,]');
+  }
+
+  initialized(contextExtensions: InitializedContextExtensions): void {
+    // find a way to remove this undefined type. There is no undefined value
+    const extensions: {
+      [key: string]: string | boolean | number | undefined;
+    } = {};
+    for (const key of Object.keys(contextExtensions)) {
+      extensions[
+        ContextExtensionsDefintion[key as InitializedContextExtensionsKey]
+      ] =
+        contextExtensions[key as InitializedContextExtensionsKey];
+    }
+
+    extensions[ContextExtensionsDefintion.sessionId] = this.sessionId;
+
+    this.duration = contextExtensions.length;
+    this.startedAt = DateTime.utc();
+
+    const data: DataPayload = {
+      context: {
+        extensions,
+      },
+      verb: {
+        display: {
+          'en-US': 'initialized',
+        },
+        id: verbDefinition.initialized,
+      },
+    };
+
+    this.send(data);
+  }
+
+  played(resultExtensions: PlayedResultExtensions): void {
+    const time: number = XAPIStatement.toFixed(resultExtensions.time);
+    this.addStartSegment(time);
+    const data: DataPayload = {
+      context: {
+        extensions: {
+          [ContextExtensionsDefintion.sessionId]: this.sessionId,
+        },
+      },
+      result: {
+        extensions: {
+          [ResultExtensionsDefinition.time]: time,
+        },
+      },
+      verb: {
+        display: {
+          'en-US': 'played',
+        },
+        id: verbDefinition.played,
+      },
+    };
+
+    this.send(data);
+  }
+
+  paused(
+    contextExtensions: PausedContextExtensions,
+    resultExtensions: PausedResultExtensions,
+  ): void {
+    const time: number = XAPIStatement.toFixed(resultExtensions.time);
+    this.addEndSegment(time);
+    const data: DataPayload = {
+      context: {
+        extensions: {
+          [ContextExtensionsDefintion.sessionId]: this.sessionId,
+        },
+      },
+      result: {
+        extensions: {
+          [ResultExtensionsDefinition.time]: time,
+        },
+      },
+      verb: {
+        display: {
+          'en-US': 'paused',
+        },
+        id: verbDefinition.paused,
+      },
+    };
+
+    data.result!.extensions[
+      ResultExtensionsDefinition.playedSegment
+    ] = this.playedSegment;
+    data.result!.extensions[
+      ResultExtensionsDefinition.progress
+    ] = XAPIStatement.toFixed(resultExtensions.time / this._duration);
+
+    if (contextExtensions.completionTreshold) {
+      data.context!.extensions[ContextExtensionsDefintion.completionTreshold] =
+        contextExtensions.completionTreshold;
+    }
+
+    this.send(data);
+  }
+
+  seeked(resultExtensions: SeekedResultExtensions): void {
+    const timeFrom: number = XAPIStatement.toFixed(resultExtensions.timeFrom);
+    const timeTo: number = XAPIStatement.toFixed(resultExtensions.timeTo);
+    this.addStartSegment(timeFrom);
```

Comment:
Should this be the other way around? `timeFrom` the end segment and `timeTo` the start segment?

Reply:
no `TimeFrom` is the time when you start seeking and `TimeTo` is the time you end seeking. So in a segment `TimeFrom` is the beginning` and `TimeTo` the end

Reply:
So this means `TimeFrom` is the end of the segment the user was previously watching and `TimeTo` the beginning of the segment they start to watch after seeking?

Or did I misunderstand something?

Reply:
> So this means TimeFrom is the end of the segment the user was previously watching

not if he was in pause.
Look the tests I made I think I cover multiple scenario possible.

Reply:
or if he start seeking before playing the video.

Reply:
> Look the tests I made I think I cover multiple scenario possible.

Sorry, since the PR is quite large I did not spend a lot of time reading the tests. The names just seemed counter-intuitive to me hence my comments :)